Build failure: qtcreator

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

`nix build .#qtcreator`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/310479557

### Relevant log output

```console
FAILED: [code=1] src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/camerageometry.cpp.o 
/nix/store/al7l1c2nwk9c9cicp3l7jfjxi1vilp93-clang-wrapper-21.1.2/bin/clang++ -DENABLE_INTERNAL_QML_RUNTIME -DIMPORT_QUICK3D_ASSETS -DNANOTRACE_ENABLED -DQMLPUPPETCOMMUNICATION_LIBRARY -DQT_CONCURRENT_LIB -D>
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/camerageometry.cpp:157:17: error: no member named 'calculateGlobalVariables' in 'QSSGRenderCamera'
  157 |         camera->calculateGlobalVariables(rect);
      |         ~~~~~~  ^
1 error generated.
[4063/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/lightgeometry.cpp.o
[4064/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/geometrybase.cpp.o
[4065/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/gridgeometry.cpp.o
[4066/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/instances/nodeinstanceclientproxy.cpp.o
[4067/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/linegeometry.cpp.o
[4068/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/mousearea3d.cpp.o
FAILED: [code=1] src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/mousearea3d.cpp.o 
/nix/store/al7l1c2nwk9c9cicp3l7jfjxi1vilp93-clang-wrapper-21.1.2/bin/clang++ -DENABLE_INTERNAL_QML_RUNTIME -DIMPORT_QUICK3D_ASSETS -DNANOTRACE_ENABLED -DQMLPUPPETCOMMUNICATION_LIBRARY -DQT_CONCURRENT_LIB -D>
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/mousearea3d.cpp:812:89: error: too few arguments to function call, single argument 'globalTransform' was not specified
  812 |             rayPos1 = rayPos0 - rayPos0.length() * DoubleVec3D(cameraNode->getDirection());
      |                                                                ~~~~~~~~~~~~~~~~~~~~~~~~ ^
/nix/store/vf2a3yv8vg48750vq8z178c9gcg0dhpi-qtquick3d-6.10.0/include/QtQuick3DRuntimeRender/6.10.0/QtQuick3DRuntimeRender/private/qssgrendernode_p.h:132:36: note: 'getDirection' declared here
  132 |     [[nodiscard]] static QVector3D getDirection(const QMatrix4x4 &globalTransform);
      |                                    ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/mousearea3d.cpp:1041:84: error: too few arguments to function call, single argument 'globalTransform' was not specified
 1041 |                 dir -= static_cast<QSSGRenderCamera *>(renderCamera)->getDirection();
      |                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
/nix/store/vf2a3yv8vg48750vq8z178c9gcg0dhpi-qtquick3d-6.10.0/include/QtQuick3DRuntimeRender/6.10.0/QtQuick3DRuntimeRender/private/qssgrendernode_p.h:132:36: note: 'getDirection' declared here
  132 |     [[nodiscard]] static QVector3D getDirection(const QMatrix4x4 &globalTransform);
      |                                    ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
[4069/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/generalhelper.cpp.o
FAILED: [code=1] src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/generalhelper.cpp.o 
/nix/store/al7l1c2nwk9c9cicp3l7jfjxi1vilp93-clang-wrapper-21.1.2/bin/clang++ -DENABLE_INTERNAL_QML_RUNTIME -DIMPORT_QUICK3D_ASSETS -DNANOTRACE_ENABLED -DQMLPUPPETCOMMUNICATION_LIBRARY -DQT_CONCURRENT_LIB -D>
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/generalhelper.cpp:365:47: error: no member named 'globalTransform' in 'QSSGRenderModel'
  365 |                         center = renderModel->globalTransform.map(bounds.center());
      |                                  ~~~~~~~~~~~  ^
1 error generated.
[4070/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/boxgeometry.cpp.o
[4071/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/lookatgeometry.cpp.o
[4072/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/selectionboxgeometry.cpp.o
FAILED: [code=1] src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet/editor3d/selectionboxgeometry.cpp.o 
/nix/store/al7l1c2nwk9c9cicp3l7jfjxi1vilp93-clang-wrapper-21.1.2/bin/clang++ -DENABLE_INTERNAL_QML_RUNTIME -DIMPORT_QUICK3D_ASSETS -DNANOTRACE_ENABLED -DQMLPUPPETCOMMUNICATION_LIBRARY -DQT_CONCURRENT_LIB -D>
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/selectionboxgeometry.cpp:199:35: error: no member named 'calculateGlobalVariables' in 'QSSGRenderNode'
  199 |                 targetRN->parent->calculateGlobalVariables();
      |                 ~~~~~~~~~~~~~~~~  ^
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/selectionboxgeometry.cpp:200:39: error: no member named 'globalTransform' in 'QSSGRenderNode'
  200 |                 m = targetRN->parent->globalTransform;
      |                     ~~~~~~~~~~~~~~~~  ^
/build/qt-creator-opensource-src-17.0.2/src/tools/qmlpuppet/qmlpuppet/editor3d/selectionboxgeometry.cpp:204:21: error: no member named 'calculateGlobalVariables' in 'QSSGRenderNode'
  204 |             rootRN->calculateGlobalVariables();
      |             ~~~~~~  ^
3 errors generated.
[4073/4303] Building CXX object src/tools/qmlpuppet/CMakeFiles/qmlpuppet.dir/qmlpuppet_autogen/mocs_compilation.cpp.o
ninja: build stopped: subcommand failed.
``
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.17.2, NixOS, 25.11 (Xantusia), 25.11.20251015.544961d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.32.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/fx3cwaih6v7sb9scmr4ckr2s47fs6say-rcsmdj802sk3w9l8m0b0bvdy0jiw2wgr-source`


### Notify maintainers

@wineee @mourogurt

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/tools/qtcreator/default.nix
+++ b/pkgs/development/tools/qtcreator/default.nix
@@ -2,6 +2,7 @@
   stdenv,
   lib,
   fetchurl,
+  fetchpatch,
   cmake,
   pkg-config,
   ninja,
@@ -38,6 +39,14 @@ stdenv.mkDerivation (finalAttrs: {
     url = "mirror://qt/official_releases/${finalAttrs.pname}/${lib.versions.majorMinor finalAttrs.version}/${finalAttrs.version}/qt-creator-opensource-src-${finalAttrs.version}.tar.xz";
     hash = "sha256-sOEY+fuJvnF2KLP5JRwpX6bfQfqLfYEhbi6tg1XlWhM=";
   };
+
+  patches = [
+    # QmlDesigner: Compile fixes for Qt 6.10 private API changes
+    (fetchpatch {
+      url = "https://github.com/qt-creator/qt-creator/commit/5a4c700ccefc76c7c531c834734e6fefa14b5364.patch";
+      hash = "sha256-BnS0HOqP5b7ZsVtuRpCK+TtoJj0yhodDuVtp+C3btIA=";
+    })
+  ];
 
   nativeBuildInputs = [
     cmake

